TS-30166 standard artifact storage schema (#211)
cqse/teamscale-jacoco-agent: agent/README.md

    - `[GET] /message` Returns the name of the currently configured commit message.
    - `[PUT] /message` Sets the commit message to the string delivered in the request body in plain text. This message
      should be used for all followup report dumps (see `teamscale-message`).
    - `[POST] /dump` Instructs the agent to dump the collected coverage.
    - `[POST] /reset` Instructs the agent to reset the collected coverage. This will discard all coverage collected in
      the current JVM session.
    - `[GET] /revision` Returns the current revision used for uploading to Teamscale.
    - `[PUT] /revision` Sets the revision to use for uploading to Teamscale. The revision must be in the request body in
      plain text.
    - `[GET] /commit` Returns the current commit used for uploading to Teamscale.
    - `[PUT] /commit` Sets the commit to use for uploading to Teamscale. The commit must be in the request body in plain
      thext in the format: branch:timestmap
- `artifactory-url`: the HTTP(S) url of the artifactory server to upload the reports to. The URL may include a subpath
  on the artifactory server, e.g. `https://artifactory.acme.com/my-repo/my/subpath`.

> - `artifactory-partition`: the partition name that is parsed later on via Teamscale Artifactory connector options.

Reviewer: I think I misread this sentence the first time. Sorry 🙈  I guess parsed was the correct term. Without the context it seemed like `pass on sth.`

- `artifactory-user` (required for artifactory): The name of an artifactory user with write access.
- `artifactory-password` (required for artifactory): The password of the user.
- `artifactory-api-key` (alternative to `artifactory-user` and `artifactory-password`) The API key for artifactory from
  a user with write access (
  c.f. [Artifactory Documentation](https://www.jfrog.com/confluence/display/JFROG/User+Profile#UserProfile-APIKey))
- `artifactory-zip-path` (optional): The path within the stored ZIP file where the reports are stored. Default is to
  store at root level. This can be used to encode e.g. a partition name that is parsed later on via Teamscale
  Artifactory connector options.
- `artifactory-path-suffix` (optional): The path within the storage location between
  the `<root-path>/uploads/<branch name>/<commit timestamp>[-<commit hash>]?/<upload partition>/<artifact type>` and the

>   uploaded artifact.

Reviewer: It is unclear to me what kind of value I should pass in as a user here. Is it `uploads`? Is it the whole pattern that is shown here? Is it something that would come even after that presented pattern?

On a related topic: We should document somewhere (CHANGELOG) maybe how the upgrade path for existing users looks like. Do we recommend to adjust the configuration in Teamscale and use the new default format? How would the user need to configure the agent to keep the old behavior (in the valid case that the customer does not want to loose all old uploads)...

Author: Regarding the second comment: Currently it's not possible to edit the default. There's a feature ticket in the backlog but it's not implemented yet.

Reviewer: @karottenreibe What is the strategy from your point of view that we should push towards? I would just want to avoid that this introduces such a breaking change that existing users cannot use any newer version of the agent without loosing all previous data.

Reviewer: @andipunz We definitely need to state somewhere what this "default path" actually is and how to configure it in Teamscale manually even though it will be the default in the next Teamscale versions. Without this information it is impossible (without reading the code) to find out how to fit those pieces together.

Author: I implemented a legacy-upload option in order to keep the new behavior. I've just checked the issues regarding the support in TS itself. They are still in dev. Should we provide somewhere in the docs (here? TS-docs) how the extraction pattern needs to be modified to use this new version?

Reviewer: Definitely yes. (See my other comment for more details)

  - `[GET] /commit` Returns the current commit used for uploading to Teamscale.
  - `[PUT] /commit` Sets the commit to use for uploading to Teamscale. The commit must be in the request body in plain thext in the format: branch:timestmap
- `sap-nwdi-applications` needed when profiling in a SAP NetWeaver Development Infrastructure. It must be a semicolon
  separated list of applications. Each application is specified as a fully qualified classname (referred to as marker
  class) and a Teamscale project alias or ID separated by a colon. The marker class must be guaranteed to be executed
  when the application is running and is unique amongst the other deployed applications.
  E.g. `com.company.app1.Main:app1alias;com.company.app2.Starter:ts-app2-id`. The coverage is uploaded to master at
  the timestamp of the last modification date of the given marker class.

### Options for the Artifactory Upload
- `artifactory-url`: the HTTP(S) url of the artifactory server to upload the reports to.
  The URL may include a subpath on the artifactory server, e.g. `https://artifactory.acme.com/my-repo/my/subpath`.

> - `artifactory-legacy-path`: option to use the upload path as defined pre v24.0.0. Use `-artifactory-legacy-path=true` to keep the old behavior. Defaults to false. For more details see [new upload schema](#the-new-standard-upload-schema), [legacy upload schema](#the-legacy-standard-upload-schema) and [migration](#migration-from-the-legacy-to-the-new-standard-upload-schema).

Author: ```suggestion
- `artifactory-legacy-path`: option to use the upload path as defined pre v24.0.0. Use `artifactory-legacy-path=true` to keep the old behavior. Defaults to false. For more details see [new upload schema](#the-new-standard-upload-schema), [legacy upload schema](#the-legacy-standard-upload-schema) and [migration](#migration-from-the-legacy-to-the-new-standard-upload-schema).
```

  See `teamscale-git-properties-jar` for details.
- `artifactory-git-properties-commit-date-format` (optional):
  The Java data pattern `git.commit.time` is encoded with in `git.properties`. Defaults to `yyyy-MM-dd'T'HH:mm:ssZ`.

### The new standard upload schema
```
/ arbitraryly / many / directories / before / the / upload / root
    / 'uploads'
    / <branch name> 
    / <commit timestamp (Java long)>[-<commit hash>]?
    / <upload partition> 
    / <artifact type> 

>     / none / or / some / internediate / directories /

Reviewer: ```suggestion
    / none / or / some / intermediate / directories /
```

    / <Zip archive file>
```

The following options must be set in the Teamscale Artifactory Connector to import reports stored with the new standard upload schema (will get the default with one of the upcoming releases):
- Path Search Pattern: `uploads/**` or `**/uploads/**` when directories before the upload root are present
- Branch Extraction Pattern: `uploads\/([^/]+)\/\d+(?:-[abcdef0-9]+)?\/[^/]+\/[^/]+\/.*`
- Timestamp Extraction Pattern: `uploads\/[^/]+\/(\d+)(?:-[abcdef0-9]+)?\/[^/]+\/[^/]+\/.*`
- Timestamp Interpretation: `timestamp:millis`
- Prefix Extraction Pattern: `uploads\/[^/]+\/\d+(?:-[abcdef0-9]+)?\/[^/]+\/([^/]+\/.*)`
- Partition Pattern (expert option): `uploads\/[^/]+\/\d+(?:-[abcdef0-9]+)?\/([^/]+)\/[^/]+\/.*`
- Analysis report mapping (expert option): `**/jacoco/**->JACOCO`

### The legacy standard upload schema
```
/ arbitraryly / many / directories / before / the / upload / root
    / <branch name> 
    / <commit timestamp (Java long)>[-<commit hash>]?
    / <Zip archive file>
```

### Migration from the legacy to the new standard upload schema

> - Add a second artifactory connector to the Teamscale project with the new pattern next to the old one.

Reviewer: ```suggestion
- Add a second artifactory connector to the Teamscale project with the new patterns next to the already existing Artifactory connector.
```
